server/src/main/scala/com/cloudera/livy/server/recovery/StateStore.scala

```diff
@@ -0,0 +1,127 @@
+
+import com.fasterxml.jackson.databind.ObjectMapper
+import com.fasterxml.jackson.module.scala.DefaultScalaModule
+
+import com.cloudera.livy.{LivyConf, Logging}
+import com.cloudera.livy.sessions.SessionKindModule
+
+trait StateStoreCompanion {
+  def create(livyConf: LivyConf): StateStore
+}
+
+protected trait JsonMapper {
+  protected val mapper = new ObjectMapper()
+    .registerModule(DefaultScalaModule)
+    .registerModule(new SessionKindModule())
+
+  def serializeToString(value: Object): String = mapper.writeValueAsString(value)
+
+  def serializeToBytes(value: Object): Array[Byte] = mapper.writeValueAsBytes(value)
+
+  def deserialize[T](json: String, valueType: Class[T]): T = mapper.readValue(json, valueType)
+
+  def deserialize[T](json: Array[Byte], valueType: Class[T]): T = mapper.readValue(json, valueType)
+}
+
+/**
+ * Interface of a key-value pair storage for state storage.
+ * It's responsible for de/serialization and retrieving/storing object.
+ * It's the low level interface used by higher level classes like SessionStore.
+ *
+ * Hardcoded to use JSON serialization for now for easier ops. Will add better serialization later.
+ */
+abstract class StateStore extends JsonMapper {
+  /**
+   * Set a key-value pair to this state store. It overwrites existing value.
+   * @throws Exception Throw when persisting the state store fails.
+   */
+  def set(key: String, value: Object): Unit
+
+  /**
+   * Get a key-value pair from this state store.
+   * @return Value if the key exists. None if the key doesn't exist.
+   * @throws Exception Throw when deserialization of the stored value fails.
+   */
+  def get[T](key: String, valueType: Class[T]): Option[T]
+
+  /**
+   * Treat keys in this state store as a directory tree and
+   * return names of the direct children of the key.
+   * @return List of names of the direct children of the key.
+   *         Empty list if the key doesn't exist or have no child.
+   */
+  def getChildren(key: String): Seq[String]
+
+  /**
+   * Remove the key from this state store. Does not throw if the key doesn't exist.
+   * @throws Exception Throw when persisting the state store fails.
+   */
+  def remove(key: String): Unit
+}
+
+/**
+ * Factory to create the store chosen in LivyConf.
+ */
+object StateStore extends Logging {
+  private[this] var stateStore: Option[StateStore] = None
+
+  def init(livyConf: LivyConf): Unit = synchronized {
+    val stateStoreName = pickStateStore(livyConf)
+    initStateStore(stateStoreName, livyConf)
+  }
+
+  def get: StateStore = {
+    assert(stateStore.isDefined, "StateStore hasn't been initialized.")
+    stateStore.get
+  }
+
+  // Return the StateStore's companion object from its class name.
+  // Using reflection here to avoid static dependency on Apache Curator.
+  // scala.reflect.runtime.universe isn't thread safe. Play safe and use Manifest.
+  private def getCompanionObject[T](name : String)(implicit man: Manifest[T]) : T =
```

Comment:
AFAIK, `Manifest` is already deprecated from 2.10, better to use ClassTag instead